### DISCUSSION TITLE
Implementation of Block and Cut Vertices finding algorithm

### DISCUSSION THREAD

**Discussion by lavkan** [networkx/networkx#6283] [Algorithms]
Hello,

I recently came across a linear time algorithm for finding blocks in a graph. I wrote the blog on the algorithm. I do not find it implemented in NetworkX. I am wondering if veteran developers here would be able to implement the algorithm. The algorithm is an adaptation of depth first search algorithm and can be found in a graph theory text. 

https://medium.com/@lavanya.kannan/finding-blocks-in-graphs-in-linear-time-2a7b43b1788d

Thanks,
Lavanya

**Comment by rossbar** [comment]
At first glance this seems pretty similar conceptually to the `bridges` function, so it might be a good fit. One question: can the articulation points/blocks be extracted directly from the DFS tree? If so, then maybe it's worth adding an example that shows how to do so with `dfs_tree`.

**Reply by lavkan** [reply]
The DFS algorithm has to be tweaked to add an edge stack and maintain a computation that determines if a node is an articulation point or not at the time of backtracking. If the node is found to be an articulation point, then the edge stack is popped accordingly to define the edge set of a block. The pseudocode and proof is beautifully written in Alan Gibbons' book on Algorithmic Graph Theory. I am attaching the  pseudocode for your convenience. 
[1.18. The depth-first search for blocks algorithm..pdf](https://github.com/networkx/networkx/files/10237585/1.18.The.depth-first.search.for.blocks.algorithm.pdf)

**Comment by dschult** [comment]
I agree with @rossbar that [bridges](https://networkx.org/documentation/stable/reference/algorithms/bridges.html) is a good place to look.  Also look at [biconnected components](https://networkx.org/documentation/stable/reference/algorithms/component.html#biconnected-components) which includes functions for edge based and node based methods as well as a function for articulation points.

**Reply by lavkan** [reply]
It is indeed implemented a function of articulation points/ bicononnected components, and the algorithm is known as Tarjan's algorithm. I think I was looking for the keyword "blocks" (same as biconnected components) and cut vertices (same as articulation points) Thanks!
